core/src/main/scala/dagr/core/tasksystem/Retry.scala

```diff
@@ -0,0 +1,137 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package dagr.core.tasksystem
+
+import java.nio.file.Files
+
+import dagr.commons.util.LazyLogging
+import dagr.core.execsystem._
+import dagr.commons.io.Io
+
+/** A trait to facilitate retry a task when it has failed. */
+trait Retry {
+  /** This method will be invoked to allow the task to perform any self-modification before retrying.
+    *
+    * @param resources the maximum system resources available to the task
+    * @param taskInfo the task execution information for the task to be retried.
+    * @return true if the task is to be retried, false otherwise.
+    */
+  def retry(resources: SystemResources, taskInfo: TaskExecutionInfo): Boolean = false
+}
+
+/**
+  * A trait to facilitate retrying a task with more memory.
+  */
+trait MemoryRetry extends Retry with FixedResources {
+  /** Given the current memory, returns the next memory to retry this task with, or None, if the task
+    * should not be retried.
+    */
+  protected def nextMemory(currentMemory: Memory): Option[Memory]
+
+  /** True if we should use the system memory as the maximum memory, false to use the heap size. */
+  protected def useSystemMemory: Boolean = true
+
+  /** Determines if this task ran out of memory, and if so, it will get the next memory value with which to be retried. */
+  protected def ranOutOfMemory(taskInfo: TaskExecutionInfo): Boolean = true
+
+  /** Tries to get the next memory value with which to run task.  It will enforce tha the task will not require more
+    * memory than in the system.
+    * @param systemResources the system resources under which this task should run.
+    * @param taskInfo the task execution information for the task to be retried.
+    * @return true if the task is to be retried, false otherwise.
+    */
+  override final def retry(systemResources: SystemResources, taskInfo: TaskExecutionInfo): Boolean = {
+    if (taskInfo.resources.memory != this.resources.memory) throw new IllegalStateException("Scheduled memory does not equal current memory")
+    if (ranOutOfMemory(taskInfo)) {
+      val maximumMemory = if (useSystemMemory) systemResources.systemMemory else systemResources.jvmMemory
+      nextMemory(this.resources.memory)
+        .filter { memory => memory <= maximumMemory }
+        .map { memory => this.requires(this.resources.cores, memory) }
+        .isDefined
+    }
+    else {
+      false
+    }
+  }
+}
+
+/**
+  * A trait for all tasks that wish to linearly increase their memory upon each retry.
+  *
+  *  By default retries until the maximum system memory is reached, and increases by the initial amount of memory required.
+  */
+trait LinearMemoryRetry extends MemoryRetry {
+  private lazy val initialMemory: Memory = this.resources.memory
+
+  /** The maximum amount of memory with which to retry. */
+  def toMemory: Memory = Memory.none
```

Comment:
Doing this will force people to override `toMemory` to make this trait usable right?  Why not either make it abstract so you can't forget to implement it, or initialize it to `Memory.infinite`, which will effectively limit it to the system resources at runtime?